File: 백준/Gold/15989. 1， 2， 3 더하기 4/1， 2， 3 더하기 4.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const t = Number(input[0])

const numbers = input.slice(1).map(Number)
const dp = Array.from({ length: 10001 }, () =>
  Array.from({ length: 4 }, () => 0),
)
dp[1][1] = 1
dp[2][1] = 1
dp[2][2] = 1
dp[3][1] = 1
dp[3][2] = 1
dp[3][3] = 1

for (let i = 4; i < 10001; i++) {
  dp[i][1] = dp[i - 1][1]
  dp[i][2] = dp[i - 2][2] + dp[i-2][1]
  dp[i][3] = dp[i - 3][1] + dp[i - 3][2] + dp[i - 3][3]
}
const answer = []
for(let num of numbers){
  answer.push(dp[num][1] + dp[num][2] + dp[num][3])
}
console.log(answer.join('\n'))

File: 백준/Gold/15989. 1， 2， 3 더하기 4/1， 2， 3 더하기 4.py
import sys
input = sys.stdin.readline

t = int(input().strip())
dp = [[0 for _ in range(3)] for _ in range(10001)]
dp[1][0] = 1
dp[2][0] = 1
dp[2][1] = 1
dp[3][0] = 1
dp[3][1] = 1
dp[3][2] = 1

for i in range(4,10001):
    dp[i][0] = dp[i - 1][0]
    dp[i][1] = dp[i - 2][0] + dp[i - 2][1]
    dp[i][2] = dp[i - 3][0] + dp[i - 3][1] + dp[i - 3][2]
for tc in range(t):
    n = int(input().strip())
    print(str(dp[n][0] + dp[n][1] + dp[n][2]))
